### ISSUE TITLE
[RFC] PCI device allocation

### ISSUE THREAD

**Issue by cpavlina** [Bareflank/hypervisor#662]
We need a simple allocator to manage PCI device geographical addresses so we can easily allocate a unique address. To serve the needs of both physical and virtual devices, it should be able to:

- Walk the physical system's PCI device tree and populate itself with all devices it finds, so it can allocate new devices not colliding with them.
- Allocate new buses and bus-bus bridges.
- Deallocate devices and bridges.

# Suggested API

```c++
struct eapis::pci::device_id
{
    bus_type bus;
    device_type device;
    func_type func;
    bus_type secondary_bus;
};

class eapis::pci::device_allocator
{
public:
    /// Create a new, empty allocator.
    device_allocator();

    device_allocator(device_allocator &&) noexcept = default;
    device_allocator &operator=(device_allocator &&) noexcept = default;

    device_allocator(const device_allocator &) = delete;
    device_allocator &operator=(const device_allocator &) = delete;

    /// Populate this allocator from the local system's devices
    /// @return true iff the population was possible without any device ID collisions. If false,
    /// the device_allocator is unmodified.
    bool populate_physical();

    /// Populate this allocator from another allocator
    /// @return true iff the population was possible without any device ID collisions. If false,
    /// the device_allocator is unmodified.
    bool populate_from(device_allocator const &other);

    /// Add a specific device to this allocator
    /// @return true iff the population was possible without a device ID collision. If false,
    /// the device_allocator is unmodified.
    bool add(device_id device);

    /// Return whether this allocator contains a given device
    bool contains(device_id device) const;

    /// Return a new device on the next available bus
    device_id allocate();

    /// Return a new device on the specified bus
    device_id allocate(bus_type bus);

    /// Create a new bus, allocating a bus-bus bridge
    device_id allocate_bridge();

    /// Create a new bus, allocating a bus-bus bridge on the specified existing bus
    device_id allocate_bridge(bus_type bus);

    /// Deallocate a device. If device is a bus-bus bridge, all devices on the secondary bus will also
    /// be deallocated.
    void deallocate(device_id device);

    /// Clear all allocated and populated devices
    void clear();

    /// Fill a vector with all devices this allocator contains
    void enumerate(std::vector<device_id> &devices) const;
};
```

# Implementation

As each bus can only contain up to 32 devices, the set of allocated devices can be managed very easily as a 32-bit bitfield so memory allocations aren't required for each device allocation. A system can contain up to 256 buses; we could either use a static array of 256 `uint32_t` bitfields to represent the entire allocator, or a `std::vector<uint32_t>` if allocation on bus creation is acceptable and preferred to a fixed 1kB-per-allocator memory consumption. I see no need for the allocator to manage function numbers as these are typically fixed within a device.

# Questions

- Should `device_allocator::populate_physical()`, `populate_from()` and `add()` throw an exception on collision instead of using a return value?

- `uint32_t[256]` or `std::vector<uint32_t>`? I'm leaning towards the latter.

- The existing PCI code uses separate `bus_type`, `device_type`, and `func_type` values, but I thought this class would work best with a single struct merging all three. We should probably be consistent, so: should I change the existing API to use this struct too, or should I change the proposed new API to not have it?

**Comment by connojd** [comment]
A couple things:

The copy constructor and copy assignment operator should be '= delete' d. Implement move constructor and assignment instead.

 I would prefer a std::array to a std::vector; I like the idea of no mallocs

I think the device information should be put into a struct, so you should update the previous API

If collision is fatal, throw an exception, if not then just return a value (possibly printing a warning as well)

**Comment by cpavlina** [comment]
Thanks, edited to reflect proposed changes. I've also renamed `allocated_device` to `device_id` since we intend to use it outside the allocator as well.

**Comment by cpavlina** [comment]
Additions to handle allocation requests when device IDs are unavailable:

```c++
class eapis::pci::device_allocator
{
public:
    /// Status of the allocator after an allocation attempt
    enum alloc_status {
        /// Device was allocated successfully.
        ALLOC_SUCCESS,

        /// A specific device ID was requested, but that ID already exists.
        ALLOC_COLLISION,

        /// A non-bridge device was requested but there is only one remaining available
        /// device ID on the bus. The last ID is left open to create a new bridge. If a
        /// specific bus is not required, the remedy for this is to allocate a new bridge and
        /// try again.
        ALLOC_BUS_ALMOST_FULL,

        /// There are no remaining devices on the bus. If a bus was specified, it is full; if no
        /// bus was specified, there are no available device IDs on any existing bus. The
        /// remedy for this is to deallocate a device first, then try again.
        ALLOC_BUS_FULL,

        /// There are no remaining devices on any bus and there are no remaining buses
        /// on the system.
        ALLOC_FULL,
    };

    /// Populate this allocator from the local system's devices
    /// @return ALLOC_SUCCESS or ALLOC_COLLISION. On collision, the allocator
    /// has not been modified.
    alloc_status populate_physical();

    /// Populate this allocator from another allocator
    /// @return ALLOC_SUCCESS or ALLOC_COLLISION. On collision, the allocator
    /// has not been modified.
    alloc_status populate_from(device_allocator const &other);

    /// Add a specific device to this allocator
    /// @return ALLOC_SUCCESS or ALLOC_COLLISION. On collision, the allocator
    /// has not been modified.
    alloc_status add(device_id device);

    /// Get a new device on the next available bus.
    /// @param device will be populated with the allocated device, or unmodified on error
    /// @return ALLOC_SUCCESS or an error code.
    alloc_status allocate(device_id &device);

    /// Get a new device on the specified bus.
    /// @param device will be populated with the allocated device, or unmodified on error
    /// @return ALLOC_SUCCESS or an error code.
    alloc_status allocate(bus_type bus, device_id &device);

    /// Create a new bus, allocating a bus-bus bridge.
    /// @param device will be populated with the allocated device, or unmodified on error
    /// @return ALLOC_SUCCESS or an error code.
    alloc_status allocate_bridge(device_id &device);

    /// Create a new bus, allocating a bus-bus bridge on the specified existing bus.
    /// @param device will be populated with the allocated device, or unmodified on error
    /// @return ALLOC_SUCCESS or an error code.
    alloc_status allocate_bridge(bus_type bus, device_id &device);
};
```

**Comment by cpavlina** [comment]
Oh right, I forgot about `enum class`. :+1:

**Comment by cpavlina** [comment]
Closed by Bareflank/extended_apis#78
